org.eclipse.scanning.malcolm.core/src/org/eclipse/scanning/malcolm/core/MalcolmDevice.java

```diff
@@ -589,48 +590,43 @@ public void eventPerformed(MalcolmEvent<MalcolmMessage> e) {
 
 	}
 	
-	public Object getAttribute(String attribute) throws MalcolmDeviceException {
-		final MalcolmMessage message = connectionDelegate.createGetMessage(attribute);
+	public <T> IDeviceAttribute<T> getAttribute(String attributeName) throws MalcolmDeviceException {
+		final MalcolmMessage message = connectionDelegate.createGetMessage(attributeName);
 		final MalcolmMessage reply   = connector.send(this, message);
 		if (reply.getType()==Type.ERROR) {
 			throw new MalcolmDeviceException("Error from Malcolm Device Connection: " + reply.getMessage());
 		}
-		return reply.getValue();
+		
+		Object result = reply.getValue();
+		if (!(result instanceof MalcolmAttribute)) {
+			throw new MalcolmDeviceException("No such attribute: " + attributeName);
+		}
+		
+		@SuppressWarnings("unchecked")
+		IDeviceAttribute<T> attribute = (IDeviceAttribute<T>) result;
+		return attribute;
 	}
 	
-	public List<MalcolmAttribute> getAllAttributes() throws MalcolmDeviceException {
-		List<MalcolmAttribute> attributeList = new LinkedList<MalcolmAttribute>();
-		String endpoint = "";
-		final MalcolmMessage message = connectionDelegate.createGetMessage(endpoint);
+	public List<IDeviceAttribute<?>> getAllAttributes() throws MalcolmDeviceException {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png 'Severity: CRITICAL') Remove usage of generic wildcard type. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS1452)